securedrop/journalist_app/api2/__init__.py

```diff
@@ -0,0 +1,137 @@
+import hashlib
+from typing import Any, Dict
+
+from flask import Blueprint, abort, json, jsonify, request
+from models import Source
+from sqlalchemy.orm import joinedload
+from werkzeug.wrappers.response import Response
+
+blp = Blueprint("api2", __name__, url_prefix="/api/v2")
+
+
+def json_version(d: dict) -> str:
+    """
+    Calculate the version (BLAKE2s digest) of the normalized JSON representation
+    of the dictionary ``d``.
+
+    We use BLAKE2s here because SHA-256 is too slow (we don't care about
+    cryptographic security) and CRC-32 is too collision-prone (we're not merely
+    checksumming for transmission integrity).
+    """
+    s = json.dumps(d, sort_keys=True)
+    b = s.encode("utf-8")
+    return hashlib.blake2s(b).hexdigest()
+
+
+@blp.get("/index")
+def index() -> Response:
+    """
+    Return the index of all sources.
+
+    If the request's ``If-None-Match`` header matches the current ETag, this
+    view MUST return HTTP 304 with an empty response.
+    """
+    sources = {}
+    for source in (
+        Source.query.options(joinedload(Source.star))
+        .options(joinedload(Source.submissions))
+        .options(joinedload(Source.replies))
+        .all()
+    ):
+        all_source_metadata = source.to_api_v2()
+        source_info: Dict[str, Any] = {
+            "version": json_version(all_source_metadata["source"]),
+            "collection": {},
+        }
+        for uuid, item in all_source_metadata["collection"].items():
+            source_info["collection"][uuid] = json_version(item)
+        sources[source.uuid] = source_info
+
+    index = {"sources": sources}
+    version = json_version(index)
+    response = jsonify(index)
+    response.set_etag(version)
+    return response.make_conditional(request)
```

Comment:
Move from specification-level docstring:

```suggestion

    # If the request's `If-None-Match` header matches the version,
    # return HTTP 304 with an empty response.
    response.set_etag(version)
    return response.make_conditional(request)
```